Feat: Composite Design Pattern for installers

tsgcpp opened the pull request:
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] No compiler errors or warnings

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## Issue Number

<!-- Referencing an issue makes the creation of CHANGELOGS for new releases much easier -->
Issue Number: #211

*Create or search an issue here: [Extenject/Issues](https://github.com/svermeulen/Extenject/issues)*

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

- No changes against current behaviours

## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Add "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller" to "OptionalExtras". They are the composite pattern of Extenject installers
- Add a document about them named "CompositeInstaller.md", so I wonder if you could also read the md file
- Add tests for "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller"
  - All asset files and prefabs are for the tests.
- Add editor scripts for "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller" to warn circular references if any circular references are found in the installers

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information
<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->
- You can see the components looks in "CompositeInstaller.md"

On which Unity version has this been tested?
--------------------------------------------
- [ ] 2020.4 LTS
- [ ] 2020.3
- [x] 2020.2
- [ ] 2020.1
- [x] 2019.4 LTS
- [ ] 2019.3
- [ ] 2019.2
- [ ] 2019.1
- [ ] 2018.4 LTS

**Scripting backend:**
- [x] Mono
- [x] IL2CPP

tsgcpp committed:
Add "CompositeMonoInstaller" and "CompositeScriptableObjectInstaller"

tsgcpp committed:
Add integration tests and test assets for composite installers

tsgcpp committed:
Add editors for composite installers

tsgcpp committed:
Add unit tests for CompositeInstallerExtensions

tsgcpp committed:
Add a documentation "CompositeInstaller.md" and images for it

tsgcpp committed:
docs: update TOC

Mathijs-Bakker commented:
@tsgcpp This is an awesome feature!
It's a lot to review, so it could take some days.
I like the tooling you've done and the circular dependency warning.

Great work!

Mathijs-Bakker commented on UnityProject/Assets/Plugins/Zenject/OptionalExtras/CompositeInstaller/Tests/IntegrationTests/Install/TestCompositeMonoInstaller/FooInjectee.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Zenject.Tests.Installers.CompositeMonoInstallers
+{
+    public class FooInjectee
```
Sorry, but could this be a typo? Did you mean FooInjector.cs?

Mathijs-Bakker replied:
My bad... this is perfect grammar X)

Mathijs-Bakker reviewed (approved):
Well written clean code and tests!

Mathijs-Bakker commented:
@tsgcpp This feature is a great addition to Zenject. 
I can merge your pr as submitted. Which means your feature will be in 'Optional Extras',
But it deserves to be standard functionality, so we could move it into into Zenject/source/...
What about it? Merge (for now) or move to the core?

tsgcpp committed:
Move composite installers scripts to "Zenject/source/..."

tsgcpp committed:
Move composite installers editor scripts to "Zenject/Source/Editor/..."

tsgcpp committed:
Move composite installers tests and to "Zenject/OptionalExtras/IntegrationTests/..."

tsgcpp committed:
Fix test names for composite installers

tsgcpp committed:
Delete unused asmdef and meta for composite installers

tsgcpp commented:
@Mathijs-Bakker I really appreciate your review and your suggestion.

> But it deserves to be standard functionality, so we could move it into into Zenject/source/...
What about it? Merge (for now) or move to the core?

If there is no problem, I would like to move them into "Zenject/source/...".
Would you tell me where to move them?

Is it ok to move them like this?
- https://github.com/tsgcpp/Extenject/compare/feature/composite-installer...tsgcpp:wip/feature/composite-installer_move_as_standard?expand=1
  - move scripts into "Zenject/source/..."
  - move tests and assets for them into "Zenject/OptionalExtras/IntegrationTests/..."
  - delete asmdef files for CompositeInstallers

Mathijs-Bakker commented:
> Is it ok to move them like this?
>   
>   * move scripts into "Zenject/source/..."
>   * move tests and assets for them into "Zenject/OptionalExtras/IntegrationTests/..."
>   * delete asmdef files for CompositeInstallers

Yes exactly.
You have all paths correct.

tsgcpp commented:
@Mathijs-Bakker OK, I've moved composite installers and their tests.
Could you please confirm my new commits?

Mathijs-Bakker commented on Documentation/CompositeInstaller.md:
```diff
@@ -0,0 +1,58 @@
+  - [FYI](#fyi)
+
+</details>
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
```
This piece of code is generated by a GitHub action. (DocToc)
These lines (1 to 14) can be moved to to under the ``## Table Of Contents`` header without causing any issues.

Mathijs-Bakker commented on Documentation/CompositeInstaller.md:
```diff
@@ -0,0 +1,58 @@
+## Table Of Contents
+- [CompositeMonoInstaller](#compositemonoinstaller)
+- [CompositeScriptableObjectInstaller](#compositescriptableobjectinstaller)
+
```
Line 19 and 20 can be removed.
When the DocToc generated lines are moved to here.

Mathijs-Bakker commented on UnityProject/Assets/Plugins/Zenject/Source/Install/CompositeScriptableObjectInstaller.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Zenject
+{
+    [CreateAssetMenu(fileName = "CompositeScriptableObjectInstaller", menuName = "Installers/CompositeScriptableObjectInstaller")]
```
This ``CreateAssetMenu`` line can be removed 
after a new ``[MenuItem("Assets/Create/Zenject/...]`` is created (in Editor/ZenMenuItems.cs)

Mathijs-Bakker reviewed (changes requested):
Cool, 
Now we decided to move this feat to Zenject's core we also want to rearrange the ``CompositeScriptableObjectInstaller`` menu item from Installers to Zenject. 

<img width="465" alt="Schermafbeelding 2021-02-07 om 12 43 12" src="https://user-images.githubusercontent.com/7645831/107146285-656a6c80-6947-11eb-8c94-d4d5707d62c0.png">

This menu is generated in ``Zenject/Source/Editor/ZenMenuItems.cs``.
Where we can add something like:
```cs
[MenuItem("Assets/Create/Zenject/Composite Scriptable Object Installer", false, 1)]
public static void CreateCompositeScriptableObjectInstaller()
{
    AddCSharpClassTemplate("Composite Scriptable Object Installer", "UntitledInstaller",
        "using UnityEngine;"
        + "\nusing Zenject;"
        // -->
}
```
For a better UX it make sense to reorder the Installer section.
So it reads like this:
- Installer
- Mono Installer
- Scriptable Object Installer
- Composite Scriptable Object Installer
        
May I ask you to take care of this?
Drop me a line for any help.

Cheers!

tsgcpp committed:
Fix "Table Of Contents" of "CompositeInstaller.md" according to other documents

tsgcpp committed:
Reorder the installer section in "ZenMenuItems"

tsgcpp committed:
Add "CreateCompositeScriptableObjectInstaller" to "ZenMenuItems"

tsgcpp committed:
Remove "CreateAssetMenu" from "CompositeScriptableObjectInstaller"

tsgcpp commented:
@Mathijs-Bakker Now, I'm editing "ZenMenuItems".

![screenshot 2021-02-07 234750](https://user-images.githubusercontent.com/19503967/107150043-2ff16d80-699f-11eb-8061-5c677a3be8e7.jpg)

I have a question.
Should I add a separator between other installers and "Composite Scriptable Object Installer" like this?

![image](https://user-images.githubusercontent.com/19503967/107150578-c030b200-69a1-11eb-9a7a-917053e064af.png)

`Installer`, `Mono Installer`, `Scriptable Object Installer` will create a "script" file, but "Composite Scriptable Object Installer" will create an "asset" file of "CompositeScriptableObjectInstaller".

Would you give me your opinion?

## WIP diff
https://github.com/tsgcpp/Extenject/compare/feature/composite-installer...tsgcpp:wip/feature/composite-installer?expand=1

Mathijs-Bakker commented:
> @Mathijs-Bakker Now, I'm editing "ZenMenuItems".
> 
> ![screenshot 2021-02-07 234750](https://user-images.githubusercontent.com/19503967/107150043-2ff16d80-699f-11eb-8061-5c677a3be8e7.jpg)
> 
> I have a question.
> Should I add a separator between other installers and "Composite Scriptable Object Installer" like this?
> 
> ![image](https://user-images.githubusercontent.com/19503967/107150578-c030b200-69a1-11eb-9a7a-917053e064af.png)
> 

Without the line separator looks the cleanest.

That was also my initial concern to reverse the order of the installers. 
- Installer
- Mono Installer
- Scriptable Object Installer
- Composite Scriptable Object Installer

It is clearer for the user to grasp the differences between the installers. Don't you think?

tsgcpp commented:
@Mathijs-Bakker Thank you for your opinion.

> It is clearer for the user to grasp the differences between the installers. Don't you think?

Yes, I think so.
OK, I will edit like this:
![screenshot 2021-02-07 234750](https://user-images.githubusercontent.com/19503967/107150043-2ff16d80-699f-11eb-8061-5c677a3be8e7.jpg)

Could you wait for a few days?
I will complete it and fix the document.

Mathijs-Bakker commented:
> Could you wait for a few days?
> I will complete it and fix the document.

No worries!
Take your time!

tsgcpp committed:
Update a step in CompositeInstaller.md to create "CompositeMonoInstaller"
Also update "CompositeScriptableObjectInstallerCreateAsset.jpg"

tsgcpp committed:
Move "ErrorTooltip" into CompositeInstallerEditorDescriptions

tsgcpp committed:
Add circular reference detection to "UnityInspectorListEditor"

tsgcpp commented:
@Mathijs-Bakker I've added  a menu item "Composite Scriptable Object Installer" and fixed "Table Of Contents" of "CompositeInstaller.md".

I also edited "UnityInspectorListEditor" to warn circular references in any contexts like this (not pushed yet):
![image](https://user-images.githubusercontent.com/19503967/107225278-f3377c00-6a5b-11eb-8722-5b884202caae.png)

Should I push this ([diff](https://github.com/tsgcpp/Extenject/compare/14e5750...fbc3f3c?expand=1))?

Mathijs-Bakker commented:
Yeah, you can push this diff to this pr.
Then we have everything in place. @tsgcpp 
👍

tsgcpp commented:
@Mathijs-Bakker I've pushed the diff.
~~Could you please confirm new commits?~~

tsgcpp commented:
@Mathijs-Bakker 
I'm sorry. Can you give me more time?
It seems that there is a bug when Extenject validation are executed.
Then, Unity freezes.

tsgcpp commented:
I found the cause of the bug.
If there is a circular references in any composite installers, "StackOverflowException" is not sometimes thrown and Unity is terminated.
I should have added a validation to composite installers before their "InstallBindings"...

I will push new commits to fix that when it is resolved.

tsgcpp committed:
Add "ValidateLeafInstallerSavedAlloc"
A validation method to avoid allocations as much as possible

tsgcpp committed:
Rename the validation method for composite to "ValidateAsComposite"

tsgcpp committed:
Add validation methods without allocations for composite installer

tsgcpp committed:
Fix composite installer validations in Editor

tsgcpp committed:
Add a validation to "CompositeMonoInstaller" before the injection

tsgcpp committed:
Add a validation to "CompositeScriptableObjectInstaller" before the injection

tsgcpp committed:
Add test assets for circular references in "CompositeMonoInstaller"

tsgcpp committed:
Add test assets for circular references in "CompositeScriptableObjectInstaller"

tsgcpp committed:
Fix the message of "ErrorTooltip" in "CompositeInstallerEditorDescriptions"

tsgcpp committed:
Fix "CompositeScriptableObjectInstallerCircularReference.jpg"

tsgcpp commented:
OK, I've resolved the bug and added assertions to composite installers before "InstallBindings" of their leaves.
If any circular references are found in composite installers, "ZenjectException" will be thrown in Runtime and "Zenject -> Validate ...".
I also fixed validation methods to reduce allocations because the validation methods became used in Runtime (No allocations up to 4 chains).

Could you please confirm new commits again? @Mathijs-Bakker

Mathijs-Bakker reviewed (approved):
@tsgcpp
Good! I see you have added the asserts.
Great!

Runned this locally (2019.LTS)... everything behaved as expected. 
Awesome work!

I will merge this pr tomorrow. 

The readme.md needs a reference to the new doc too. And a brief explanation about this feature.
I can append a section to it.

:star: :star: :star: Awesome work! :star: :star: :star:

Mathijs-Bakker commented:
Merged.

tsgcpp commented:
I really appreciate your supports! @Mathijs-Bakker 
I hope this contribution will help the developers.
